File: solutions/elixir/dnd-character/1/lib/dnd_character.ex
defmodule DndCharacter do
  @type t :: %__MODULE__{
          strength: pos_integer(),
          dexterity: pos_integer(),
          constitution: pos_integer(),
          intelligence: pos_integer(),
          wisdom: pos_integer(),
          charisma: pos_integer(),
          hitpoints: pos_integer()
        }
  @sides_on_die 6

  defstruct ~w[strength dexterity constitution intelligence wisdom charisma hitpoints]a

  @spec modifier(pos_integer()) :: integer()
  def modifier(score),
    do: Kernel.trunc(score / 2) - 5

  @spec ability :: pos_integer()
  def ability(),
    do:
      [roll_die(), roll_die(), roll_die(), roll_die()]
      |> drop_lowest()
      |> Enum.sum()

  defp roll_die(),
    do: :rand.uniform(@sides_on_die)

  defp drop_lowest(dice_values) do
    minimum_value = Enum.min(dice_values)
    minimum_index = Enum.find_index(dice_values, &(&1 === minimum_value))
    {_, new_dice_values} = List.pop_at(dice_values, minimum_index)

    new_dice_values
  end

  @spec character :: t()
  def character do
    constitution = ability()

    %DndCharacter{
      strength: ability(),
      dexterity: ability(),
      constitution: constitution,
      intelligence: ability(),
      wisdom: ability(),
      charisma: ability(),
      hitpoints: 10 + modifier(constitution)
    }
  end
end

File: solutions/java/dnd-character/1/src/main/java/DnDCharacter.java
import java.util.List;

class DnDCharacter {
	private static final int HIT_POINT_BASE = 10;

	private final Dice dice;

	private final int strength;
	private final int dexterity;
	private final int constitution;
	private final int intelligence;
	private final int wisdom;
	private final int charisma;
	private final int hitPoints;

	private DnDCharacter(Dice dice) {
		this.dice = dice;

		this.strength = ability(dice.roll());
		this.dexterity = ability(dice.roll());
		this.constitution = ability(dice.roll());
		this.intelligence = ability(dice.roll());
		this.wisdom = ability(dice.roll());
		this.charisma = ability(dice.roll());

		this.hitPoints = HIT_POINT_BASE + modifier(this.constitution);
	}

	public DnDCharacter() {
		this(new Dice());
	}

	private static final int ABILITY_DICE_TO_KEEP = 3;
	private static final int MODIFIER_BASELINE = 10;
	private static final int MODIFIER_DIVISOR = 2;

	int ability(List<Integer> scores) {
		return scores.stream()
				       .sorted((a, b) -> b - a)  // descending
				       .limit(ABILITY_DICE_TO_KEEP)
				       .mapToInt(Integer::intValue)
				       .sum();
	}

	List<Integer> rollDice() {
		return dice.roll();
	}

	int modifier(int score) {
		return Math.floorDiv(score - MODIFIER_BASELINE, MODIFIER_DIVISOR);
	}

	int getStrength()     { return strength; }
	int getDexterity()    { return dexterity; }
	int getConstitution() { return constitution; }
	int getIntelligence() { return intelligence; }
	int getWisdom()       { return wisdom; }
	int getCharisma()     { return charisma; }
	int getHitpoints()    { return hitPoints; }
}
